osu.Game.Tests/Visual/Gameplay/TestSceneStoryboardSamplePlayback.cs

```diff
@@ -47,26 +60,86 @@ public void TestStoryboardSamplesStopDuringPause()
         [Test]
         public void TestStoryboardSamplesStopOnSkip()
         {
-            checkForFirstSamplePlayback();
+            createPlayerTest(true);
 
-            AddStep("skip intro", () => InputManager.Key(osuTK.Input.Key.Space));
             AddAssert("all storyboard samples stopped immediately", () => allStoryboardSamples.All(sound => !sound.IsPlaying));
 
             AddUntilStep("any storyboard samples playing after skip", () => allStoryboardSamples.Any(sound => sound.IsPlaying));
         }
 
-        private void checkForFirstSamplePlayback()
+        [TestCase(typeof(OsuModDoubleTime), 1.5)]
+        [TestCase(typeof(OsuModDoubleTime), 2)]
+        [TestCase(typeof(OsuModHalfTime), 0.75)]
+        [TestCase(typeof(OsuModHalfTime), 0.5)]
+        public void TestStoryboardSamplesPlaybackWithRateAdjustMods(Type expectedMod, double expectedRate)
+        {
+            AddStep("setup mod", () =>
+            {
+                ModRateAdjust testedMod = (ModRateAdjust)Activator.CreateInstance(expectedMod).AsNonNull();
+                testedMod.SpeedChange.Value = expectedRate;
+                storyboardMods = new[] { testedMod };
+            });
+
+            createPlayerTest(true);
+
+            AddAssert("sample playback rate matches mod rates", () => allStoryboardSamples.All(sound =>
+            {
+                return sound.ChildrenOfType<DrawableSample>().All(s => s.AggregateFrequency.Value == expectedRate);
+            }));
```

Comment:
I find rewriting this test in this diff in this manner a bit dubious, especially that it's 99% going to break once the storyboard samples become pooled. But sure.

Reply:
Well it's not like there's any other way to approach this, I think?

Reply:
The previous test setup was working perfectly fine, was it not? There may be a point that it was testing at too low a level (i.e. wasn't testing integration), which caused the regression to slip by, but there was definitely a way to keep it where it was.

Reply:
Are you referring to this ensuring all `DrawableSample`s of a `DrawableStoryboardSample` have the same frequency rather than any? that's the only difference I'm seeing between previous and current code:
https://github.com/ppy/osu/blob/c197e27e33bc77da7b96990dc10d2677eda38cc4/osu.Game.Tests/Gameplay/TestSceneStoryboardSamples.cs#L168-L171

https://github.com/ppy/osu/blob/cbb8dc28914a4c7a8cb636ea5b58534869eea934/osu.Game.Tests/Visual/Gameplay/TestSceneStoryboardSamplePlayback.cs#L85-L88

Reply:
The `.All()` call can silently fail if there aren't any samples in the `DrawableSample`, because `.All()` returns true for empty collections. Which may become a concern with pooling.

It's fine for now, but it may be a decent idea to add a safety ensuring that at least one sample exists here.

Reply:
Right, good catch.